> 在arm64平台有下面的编译错误 需要解决
> 
> drivers/firmware/efi/cper.c: In function ‘cper_print_proc_generic’: drivers/firmware/efi/cper.c:235:13: error: ‘boot_cpu_data’ undeclared (first use in this function); did you mean ‘boot_cpu_init’? 235 | if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN || | ^~~~~~~~~~~~~ | boot_cpu_init drivers/firmware/efi/cper.c:235:13: note: each undeclared identifier is reported only once for each function it appears in CC [M] drivers/net/wireless/realtek/rtl8xxxu/rtl8xxxu_8710b.o drivers/firmware/efi/cper.c:235:41: error: ‘X86_VENDOR_ZHAOXIN’ undeclared (first use in this function); did you mean ‘PCI_VENDOR_ID_ZHAOXIN’? 235 | if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN || | ^~~~~~~~~~~~~~~~~~ | PCI_VENDOR_ID_ZHAOXIN drivers/firmware/efi/cper.c:236:41: error: ‘X86_VENDOR_CENTAUR’ undeclared (first use in this function); did you mean ‘X86_VENDOR_ANY’? 236 | boot_cpu_data.x86_vendor == X86_VENDOR_CENTAUR) | ^~~~~~~~~~~~~~~~~~ | X86_VENDOR_ANY CC drivers/clocksource/arm_arch_timer.o make[5]: *** [scripts/Makefile.build:243：drivers/firmware/efi/cper.o] 错误 1 make[4]: *** [scripts/Makefile.build:480：drivers/firmware/efi] 错误 2 make[3]: *** [scripts/Makefile.build:480：drivers/firmware] 错误 2

感谢指正。新版中已经处理了这个问题。

--- a/drivers/firmware/efi/cper.c
+++ b/drivers/firmware/efi/cper.c
@@ -141,6 +141,54 @@ static const char * const proc_flag_strs[] = {
 	"corrected",
 };
 
+static const char * const zdi_zpi_err_type_strs[] = {
+	"No Error",
+	"Training Error Status (PHY)",
+	"Data Link Protocol Error Status (DLL)",
+	"Surprise Down Error Status",
+	"Flow Control Protocol Error Status (TL)",
+	"Receiver Overflow Status (TL)",
+	"Receiver Error Status (PHY)",
+	"Bad TLP Status (DLL)",
+	"Bad Data Link Layer Packet (DLLP) Status (DLL)",
+	"REPLAY_NUM Rollover Status (DLL)",
+	"Replay Timer Timeout Status (DLL)",
+	"X16 Link Width Unreliable Status",
+	"ZPI X8 Link Width Unreliable Status",
+	"ZPI X4 Link Width Unreliable Status",
+	"ZPI X2 Link Width Unreliable Status",
+	"ZPI Gen3 Link Speed Unreliable Status",
+	"ZPI Gen2 Link Speed Unreliable Status",
+	"ZDI Gen3 Link Speed Unreliable Status",
+	"ZDI Gen4 Link Speed Unreliable Status",
+};
+
+const char *cper_zdi_zpi_err_type_str(unsigned int etype)
+{
+	return etype < ARRAY_SIZE(zdi_zpi_err_type_strs) ?
+			zdi_zpi_err_type_strs[etype] : "unknown error";
+}
+EXPORT_SYMBOL_GPL(cper_zdi_zpi_err_type_str);
+
+static void __maybe_unused
+cper_print_proc_generic_zdi_zpi(const char *pfx, const struct cper_sec_proc_generic *zdi_zpi)
+{
+	u8 etype = zdi_zpi->responder_id;
+
+	if ((zdi_zpi->requestor_id & 0xff) == 7) {
+		pr_info("%s general processor error(zpi error)\n", pfx);
+	} else if ((zdi_zpi->requestor_id & 0xff) == 6) {
+		pr_info("%s general processor error(zdi error)\n", pfx);
+	} else {
+		pr_info("%s general processor error(unknown error)\n", pfx);
+		return;
+	}
+	pr_info("%s bus number %llx device number %llx function number 0\n", pfx,
+		((zdi_zpi->requestor_id)>>8) & 0xff, zdi_zpi->requestor_id & 0xff);
+	pr_info("%s apic id %lld error_type:  %s\n", pfx, zdi_zpi->proc_id,
+		cper_zdi_zpi_err_type_str(etype));
+}
+
 static void cper_print_proc_generic(const char *pfx,
 				    const struct cper_sec_proc_generic *proc)
 {
@@ -184,6 +232,11 @@ static void cper_print_proc_generic(const char *pfx,
 		       pfx, proc->responder_id);
 	if (proc->validation_bits & CPER_PROC_VALID_IP)
 		printk("%s""IP: 0x%016llx\n", pfx, proc->ip);
+#if IS_ENABLED(CONFIG_X86)
+	if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN ||
+	    boot_cpu_data.x86_vendor == X86_VENDOR_CENTAUR)
+		cper_print_proc_generic_zdi_zpi(pfx, proc);
+#endif
 }
 
 static const char * const mem_err_type_strs[] = {

--- a/include/linux/cper.h
+++ b/include/linux/cper.h
@@ -578,4 +578,5 @@ void cper_estatus_print(const char *pfx,
 int cper_estatus_check_header(const struct acpi_hest_generic_status *estatus);
 int cper_estatus_check(const struct acpi_hest_generic_status *estatus);
 
+const char *cper_zdi_zpi_err_type_str(unsigned int etype);
 #endif

